// @lifi/sdk/src/_cjs/core/Solana/SolanaStepExecutor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SolanaStepExecutor = void 0;
const web3_js_1 = require("@solana/web3.js");
const bs58_1 = require("bs58");
const config_js_1 = require("../../config.js");
const constants_js_1 = require("../../errors/constants.js");
const errors_js_1 = require("../../errors/errors.js");
const api_js_1 = require("../../services/api.js");
const base64ToUint8Array_js_1 = require("../../utils/base64ToUint8Array.js");
const index_js_1 = require("../../utils/index.js");
const BaseStepExecutor_js_1 = require("../BaseStepExecutor.js");
const checkBalance_js_1 = require("../checkBalance.js");
const processMessages_js_1 = require("../processMessages.js");
const stepComparison_js_1 = require("../stepComparison.js");
const utils_js_1 = require("../utils.js");
const waitForReceivingTransaction_js_1 = require("../waitForReceivingTransaction.js");
const connection_js_1 = require("./connection.js");
const parseSolanaErrors_js_1 = require("./parseSolanaErrors.js");
class SolanaStepExecutor extends BaseStepExecutor_js_1.BaseStepExecutor {
    constructor(options) {
        super(options);
        Object.defineProperty(this, "walletAdapter", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "checkWalletAdapter", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: (step) => {
                if (this.walletAdapter.publicKey.toString() !== step.action.fromAddress) {
                    throw new errors_js_1.TransactionError(constants_js_1.LiFiErrorCode.WalletChangedDuringExecution, 'The wallet address that requested the quote does not match the wallet address attempting to sign the transaction.');
                }
            }
        });
        Object.defineProperty(this, "executeStep", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: async (step) => {
                step.execution = this.statusManager.initExecutionObject(step);
                const fromChain = await config_js_1.config.getChainById(step.action.fromChainId);
                const toChain = await config_js_1.config.getChainById(step.action.toChainId);
                const isBridgeExecution = fromChain.id !== toChain.id;
                const currentProcessType = isBridgeExecution ? 'CROSS_CHAIN' : 'SWAP';
                let process = this.statusManager.findOrCreateProcess({
                    step,
                    type: currentProcessType,
                });
                if (process.status !== 'DONE') {
                    try {
                        const connection = await (0, connection_js_1.getSolanaConnection)();
                        process = this.statusManager.updateProcess(step, process.type, 'STARTED');
                        await (0, checkBalance_js_1.checkBalance)(this.walletAdapter.publicKey.toString(), step);
                        if (!step.transactionRequest) {
                            const { execution, ...stepBase } = step;
                            const updatedStep = await (0, api_js_1.getStepTransaction)(stepBase);
                            const comparedStep = await (0, stepComparison_js_1.stepComparison)(this.statusManager, step, updatedStep, this.allowUserInteraction, this.executionOptions);
                            step = {
                                ...comparedStep,
                                execution: step.execution,
                            };
                        }
                        if (!step.transactionRequest?.data) {
                            throw new errors_js_1.TransactionError(constants_js_1.LiFiErrorCode.TransactionUnprepared, 'Unable to prepare transaction.');
                        }
                        process = this.statusManager.updateProcess(step, process.type, 'ACTION_REQUIRED');
                        if (!this.allowUserInteraction) {
                            return step;
                        }
                        let transactionRequest = {
                            data: step.transactionRequest.data,
                        };
                        const blockhashResult = await connection.getLatestBlockhash({
                            commitment: 'confirmed',
                        });
                        if (this.executionOptions?.updateTransactionRequestHook) {
                            const customizedTransactionRequest = await this.executionOptions.updateTransactionRequestHook({
                                requestType: 'transaction',
                                ...transactionRequest,
                            });
                            transactionRequest = {
                                ...transactionRequest,
                                ...customizedTransactionRequest,
                            };
                        }
                        if (!transactionRequest.data) {
                            throw new errors_js_1.TransactionError(constants_js_1.LiFiErrorCode.TransactionUnprepared, 'Unable to prepare transaction.');
                        }
                        const versionedTransaction = web3_js_1.VersionedTransaction.deserialize((0, base64ToUint8Array_js_1.base64ToUint8Array)(transactionRequest.data));
                        this.checkWalletAdapter(step);
                        const signedTxPromise = this.walletAdapter.signTransaction(versionedTransaction);
                        const signedTx = await Promise.race([
                            signedTxPromise,
                            (0, utils_js_1.sleep)(120_000),
                        ]);
                        if (!signedTx) {
                            throw new errors_js_1.TransactionError(constants_js_1.LiFiErrorCode.TransactionExpired, 'Transaction has expired: blockhash is no longer recent enough.');
                        }
                        process = this.statusManager.updateProcess(step, process.type, 'PENDING');
                        const simulationResult = await connection.simulateTransaction(signedTx, {
                            commitment: 'processed',
                            replaceRecentBlockhash: true,
                        });
                        if (simulationResult.value.err) {
                            throw new errors_js_1.TransactionError(constants_js_1.LiFiErrorCode.TransactionSimulationFailed, 'Transaction simulation failed');
                        }
                        const txSignature = bs58_1.default.encode(signedTx.signatures[0]);
                        const lastValidBlockHeight = blockhashResult.lastValidBlockHeight - 150;
                        const abortController = new AbortController();
                        const confirmTransactionPromise = connection
                            .confirmTransaction({
                            signature: txSignature,
                            blockhash: blockhashResult.blockhash,
                            lastValidBlockHeight: lastValidBlockHeight,
                            abortSignal: abortController.signal,
                        }, 'confirmed')
                            .then((result) => result.value);
                        let confirmedTx = null;
                        let blockHeight = await connection.getBlockHeight();
                        const rawTransactionOptions = {
                            skipPreflight: true,
                            maxRetries: 0,
                            preflightCommitment: 'confirmed',
                        };
                        const signedTxSerialized = signedTx.serialize();
                        while (!confirmedTx && blockHeight < lastValidBlockHeight) {
                            await connection.sendRawTransaction(signedTxSerialized, rawTransactionOptions);
                            confirmedTx = await Promise.race([
                                confirmTransactionPromise,
                                (0, utils_js_1.sleep)(1000),
                            ]);
                            if (confirmedTx) {
                                break;
                            }
                            blockHeight = await connection.getBlockHeight();
                        }
                        abortController.abort();
                        if (!confirmedTx) {
                            throw new errors_js_1.TransactionError(constants_js_1.LiFiErrorCode.TransactionExpired, 'Transaction has expired: The block height has exceeded the maximum allowed limit.');
                        }
                        if (confirmedTx?.err) {
                            const reason = typeof confirmedTx.err === 'object'
                                ? JSON.stringify(confirmedTx.err)
                                : confirmedTx.err;
                            throw new errors_js_1.TransactionError(constants_js_1.LiFiErrorCode.TransactionFailed, `Transaction failed: ${reason}`);
                        }
                        process = this.statusManager.updateProcess(step, process.type, 'PENDING', {
                            txHash: txSignature,
                            txLink: `${fromChain.metamask.blockExplorerUrls[0]}tx/${txSignature}`,
                        });
                        if (isBridgeExecution) {
                            process = this.statusManager.updateProcess(step, process.type, 'DONE');
                        }
                    }
                    catch (e) {
                        const error = await (0, parseSolanaErrors_js_1.parseSolanaErrors)(e, step, process);
                        process = this.statusManager.updateProcess(step, process.type, 'FAILED', {
                            error: {
                                message: error.cause.message,
                                code: error.code,
                            },
                        });
                        this.statusManager.updateExecution(step, 'FAILED');
                        throw error;
                    }
                }
                const processTxHash = process.txHash;
                if (isBridgeExecution) {
                    process = this.statusManager.findOrCreateProcess({
                        step,
                        type: 'RECEIVING_CHAIN',
                        status: 'PENDING',
                        chainId: toChain.id,
                    });
                }
                let statusResponse;
                try {
                    if (!processTxHash) {
                        throw new Error('Transaction hash is undefined.');
                    }
                    statusResponse = (await (0, waitForReceivingTransaction_js_1.waitForReceivingTransaction)(processTxHash, this.statusManager, process.type, step));
                    const statusReceiving = statusResponse.receiving;
                    process = this.statusManager.updateProcess(step, process.type, 'DONE', {
                        substatus: statusResponse.substatus,
                        substatusMessage: statusResponse.substatusMessage ||
                            (0, processMessages_js_1.getSubstatusMessage)(statusResponse.status, statusResponse.substatus),
                        txHash: statusReceiving?.txHash,
                        txLink: `${toChain.metamask.blockExplorerUrls[0]}tx/${statusReceiving?.txHash}`,
                    });
                    this.statusManager.updateExecution(step, 'DONE', {
                        fromAmount: statusResponse.sending.amount,
                        toAmount: statusReceiving?.amount,
                        toToken: statusReceiving?.token,
                        gasCosts: [
                            {
                                amount: statusResponse.sending.gasAmount,
                                amountUSD: statusResponse.sending.gasAmountUSD,
                                token: statusResponse.sending.gasToken,
                                estimate: statusResponse.sending.gasUsed,
                                limit: statusResponse.sending.gasUsed,
                                price: statusResponse.sending.gasPrice,
                                type: 'SEND',
                            },
                        ],
                    });
                }
                catch (e) {
                    const htmlMessage = await (0, index_js_1.getTransactionFailedMessage)(step, process.txLink);
                    process = this.statusManager.updateProcess(step, process.type, 'FAILED', {
                        error: {
                            code: constants_js_1.LiFiErrorCode.TransactionFailed,
                            message: 'Failed while waiting for receiving chain.',
                            htmlMessage,
                        },
                    });
                    this.statusManager.updateExecution(step, 'FAILED');
                    console.warn(e);
                    throw e;
                }
                return step;
            }
        });
        this.walletAdapter = options.walletAdapter;
    }
}
exports.SolanaStepExecutor = SolanaStepExecutor;
